src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Diagnostics/BindingDiagnostics.cs

```diff
@@ -0,0 +1,122 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+//
+// Description:
+//  Binding diagnostics API
+//
+
+using System.Collections.Generic;
+
+namespace System.Windows.Diagnostics
+{
+    /// <summary>
+    /// Provides a notification infrastructure for listening to binding failure events.
+    /// </summary>
+    /// <remarks>
+    /// This type supports the .NET Framework infrastructure and is not intended to be used directly
+    /// from application code.
+    /// </remarks>
+    public static class BindingDiagnostics
+    {
+        internal static bool IsEnabled { get; private set; }
+
+        private static event EventHandler<BindingFailedEventArgs> bindingFailed;
+        private static List<BindingFailedEventArgs> pendingEvents;
+        private static object pendingEventsLock;
+        private const int maxPendingEvents = 2000;
+
+        static BindingDiagnostics()
+        {
+            BindingDiagnostics.IsEnabled = VisualDiagnostics.IsEnabled && VisualDiagnostics.IsEnvironmentVariableSet(null, XamlSourceInfoHelper.XamlSourceInfoEnvironmentVariable);
+
+            if (BindingDiagnostics.IsEnabled)
+            {
+                // Listeners may miss the initial set of binding failures, so cache events until the first listener attaches.
+                // Normally there will only be one listener added soon after the process starts,
+                // and it will want to know about any binding failures that already happened.
+
+                BindingDiagnostics.pendingEvents = new List<BindingFailedEventArgs>();
+                BindingDiagnostics.pendingEventsLock = new object();
+            }
+        }
+
+        /// <summary>
+        /// Handlers of this event should return control to WPF quickly, and not cache BindingFailedEventArgs for future use.
+        /// </summary>
+        public static event EventHandler<BindingFailedEventArgs> BindingFailed
+        {
+            add
+            {
+                if (BindingDiagnostics.IsEnabled)
+                {
+                    BindingDiagnostics.bindingFailed += value;
+                    BindingDiagnostics.FlushPendingBindingFailedEvents();
+                }
+            }
+
+            remove
+            {
+                BindingDiagnostics.bindingFailed -= value;
+            }
+        }
+
+        /// <summary>
+        /// Flushes all cached binding failure events and stops any further events from being cached.
+        /// </summary>
+        private static void FlushPendingBindingFailedEvents()
+        {
+            if (BindingDiagnostics.pendingEvents != null)
+            {
+                BindingFailedEventArgs[] pendingEventsCopy = null;
+
+                lock (BindingDiagnostics.pendingEventsLock)
+                {
+                    pendingEventsCopy = BindingDiagnostics.pendingEvents?.ToArray();
```

Comment:
Can we not do call ToArray ?

Because the BindingDiagnostics.pendingEvents will never change

Reply:
Makes sense, thanks